pkg/controllers/addons_controller.go

```diff
@@ -56,151 +51,74 @@ func NewReconciler(client client.Client, logger logr.Logger,
-		config, err := clientcmd.BuildConfigFromFlags("", kubeConfig)
-		if err != nil {
-			return nil, err
-		}
-
-		// create the clientset
-		clientset, err := kubernetes.NewForConfig(config)
-		if err != nil {
-			return nil, err
-		}
-
-		return clientset, nil
-	}
-	// creates the in-cluster config
-	config, err := rest.InClusterConfig()
-	if err != nil {
-		panic(err.Error())
-	}
-	// creates the clientset
-	clientset, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		panic(err.Error())
+func processFailed(l layers.Layer) {
+	if l.GetStatus() == kraanv1alpha1.FailedCondition {
+		return
+		// Reset failed status if it was set more than 'interval' duration ago
+		// Use previous condition to detect what to set status to
 	}
-
-	return clientset, nil
-}
-
-func logBadStatusError(log logr.Logger, status string) {
-	utils.Log(log, 3, 1, fmt.Sprintf("invalid AddonLayer status: %s, ignoring", status))
 }
 
-func processEmptyStatus(l *layers.Layer) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", "")
-	l.StatusPrunePending()
-	l.SetRequeue()
-	utils.Log(l.GetLogger(), 2, 1, "processed",
-		"Previous Status", "",
-		"Status", l.GetFullStatus(),
-		"Spec:", l.GetSpec())
-}
+func processAddonLayer(l layers.Layer) error { // nolint:gocyclo // ok
+	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", l.GetStatus())
 
-func processDeployed(l *layers.Layer) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.DeployedCondition)
-	if !l.IsVersionCurrent() {
-		// Version has changed set to prune pending.
-		l.StatusPrunePending()
-		l.SetRequeue()
+	if l.IsHold() {
+		l.SetHold()
+		return nil
 	}
-	utils.Log(l.GetLogger(), 2, 1, "processed",
-		"Previous Status", kraanv1alpha1.DeployedCondition,
-		"Status", l.GetFullStatus(),
-		"Spec:", l.GetSpec())
-}
 
-func processPrunePending(l *layers.Layer) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.PrunePendingCondition)
+	processFailed(l)
+
 	if !l.CheckK8sVersion() {
-		l.SetRequeue()
+		l.SetStatusK8sVersion()
 		l.SetDelayed()
-		return
+		return nil
 	}
-	l.StatusPrune()
-	// Thinking about this, implementing a reliable reverse DependsOn is non trivial so suggest
-	// that for now we just transition to Prune state and revist waiting for layers that depend on
-	// this layer later.
-	utils.Log(l.GetLogger(), 2, 1, "processed",
-		"Previous Status", kraanv1alpha1.PrunePendingCondition,
-		"Status", l.GetFullStatus(),
-		"Spec:", l.GetSpec())
-}
 
-func processPrune(l *layers.Layer, r *AddonsLayerReconciler) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.PruneCondition)
-	l.StatusPruning()
-	if err := r.update(l.GetContext(), l.GetLogger(), l.GetAddonsLayer()); err != nil {
-		l.SetRequeue()
-		return
+	if l.IsPruningRequired() {
+		l.SetStatusPruning()
+		if err := l.Prune(); err != nil {
+			return err
+		}
+		if err := l.SetAllPrunePending(); err != nil {
+			return err
+		}
+		l.SetDelayed()
```

Comment:
Whenever we change anything we set the updated and requeue flags so that it updates the resource and requeues it for immediate reprocessing. However if the action we have just taken is likely to take some time to complete or we are waiting on an external event (like k8s version update, another AddonsLayer to deploy/prune or a HR deletion/deployment) we don't want to process it again immediately we want to wait for a period to prevent the controller constantly consuming resources doing checks that are unlikely to have a positive outcome. In the case of waiting for pruning or deployment of our HRs or other AddonsLayers to change status we will detect that because we will watch those events and can trigger reprocess of this AddonsLayer sooner by updating it or triggering reprocessing in some other way (as discussed need to determine if this is possible).